src/coreclr/jit/runtimelookup.cpp

```diff
@@ -0,0 +1,461 @@
+                    {
+                        signature = (void*)vnStore->CoercedConstantValue<ssize_t>(sigTree->gtVNPair.GetLiberal());
+                    }
+                    else
+                    {
+                        // Technically, it is possible (e.g. it was CSE'd and then VN was erased), but for Debug mode we
+                        // want to catch such cases as we really don't want to emit just a fallback call - it's too slow
+                        assert(!"can't restore signature argument value");
+                        continue;
+                    }
+                }
+                assert(signature != nullptr);
+
+                CORINFO_RUNTIME_LOOKUP runtimeLookup = {};
+                const bool             lookupFound   = GetSignatureToLookupInfoMap()->Lookup(signature, &runtimeLookup);
+                assert(lookupFound);
+
+                const bool needsSizeCheck = runtimeLookup.sizeOffset != CORINFO_NO_SIZE_CHECK;
+                if (needsSizeCheck)
+                {
+                    JITDUMP("dynamic expansion, needs size check.\n")
+                }
+
+                DebugInfo debugInfo = stmt->GetDebugInfo();
+
+                assert(runtimeLookup.indirections != 0);
+                assert(runtimeLookup.testForNull);
+
+                BasicBlockFlags originalFlags = block->bbFlags;
+                BasicBlock*     prevBb        = block;
+
+                if (prevStmt == nullptr || opts.OptimizationDisabled())
+                {
+                    JITDUMP("Splitting " FMT_BB " at the beginning.\n", prevBb->bbNum)
+                    block = fgSplitBlockAtBeginning(prevBb);
+                }
+                else
+                {
+                    JITDUMP("Splitting " FMT_BB " after statement " FMT_STMT "\n", prevBb->bbNum, prevStmt->GetID())
+                    block = fgSplitBlockAfterStatement(prevBb, prevStmt);
+                }
+
+                // We split a block, possibly, in the middle - we need to propagate some flags
+                prevBb->bbFlags =
+                    originalFlags & (~(BBF_SPLIT_LOST | BBF_LOOP_PREHEADER | BBF_RETLESS_CALL) | BBF_GC_SAFE_POINT);
+                block->bbFlags |= originalFlags & (BBF_SPLIT_GAINED | BBF_IMPORTED | BBF_GC_SAFE_POINT |
+                                                   BBF_LOOP_PREHEADER | BBF_RETLESS_CALL);
+
+                // Define a local for the result
+                const unsigned rtLookupLclNum   = lvaGrabTemp(true DEBUGARG("runtime lookup"));
+                lvaTable[rtLookupLclNum].lvType = TYP_I_IMPL;
+                GenTreeLclVar* rtLookupLcl      = gtNewLclvNode(rtLookupLclNum, call->TypeGet());
+
+                // Save expression to a local and append as the last statement in prevBb
+                auto spillExpr = [&](GenTree* expr) -> GenTree* {
+                    if (expr->OperIs(GT_LCL_VAR))
+                    {
+                        return gtClone(expr);
+                    }
```

Comment:
This looks dangerous given that we are reordering these trees with ones before the runtime lookup. There could be interference.